File: Javascript/LinkedList/206_Reverse_Linked_List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
    if (!head) return head;

    let start = head;
    let prev = head;
    let cur = head.next;

    while (cur) {
        const temp = cur.next;
        start.next = cur.next
        cur.next = prev;

        prev = cur;
        cur = temp;
    }

    return prev;
};

File: Problem-lists/Algorithm_Patterns/Linked_List/206_Reverse_Linked_List.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
    - Given an integer array nums representing the amount of money of each house
    - Return the maximum amount of money you can rob tonight without alerting the police.

    Time: O(N), Space complexity: O(1)
*/

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

ListNode* reverseList(ListNode* head) {
    if (!head || !head->next) return head;
    ListNode* cur = head;
    ListNode* next = head->next;

    while (next) {
        ListNode* temp = next->next;
        next->next = cur;
        cur = next;
        next = temp;
    }
    head->next = nullptr;

    return cur;
}